state/protocol/badger/mutator.go

```diff
@@ -404,32 +399,26 @@ func (m *FollowerState) lastSealed(candidate *flow.Block) (*flow.Seal, error) {
 	return last, nil
 }
 
-// insert stores the candidate block in the data base. The
-// `candidate` block _must be valid_ (otherwise, the state will be corrupted).
+// insert stores the candidate block in the database.
+// The `candidate` block _must be valid_ (otherwise, the state will be corrupted).
+// dbUpdates contains other database operations which must be applied atomically
+// with inserting the block.
 func (m *FollowerState) insert(ctx context.Context, candidate *flow.Block, last *flow.Seal) error {
 
 	span, _ := m.tracer.StartSpanFromContext(ctx, trace.ProtoStateMutatorExtendDBInsert)
 	defer span.Finish()
 
 	blockID := candidate.ID()
 
-	// SIXTH: epoch transitions and service events
-	//    (i) Determine protocol state for block's _current_ Epoch.
-	//        As we don't have slashing yet, the protocol state is fully
-	//        determined by the Epoch Preparation events.
-	//   (ii) Determine protocol state for block's _next_ Epoch.
-	//        In case any of the payload seals includes system events,
-	//        we need to check if they are valid and must apply them
-	//        to the protocol state as needed.
-	ops, err := m.handleServiceEvents(candidate)
+	// apply any state changes from service events sealed by this block's parent
+	dbUpdates, insertingBlockTriggersEpochFallback, err := m.handleEpochServiceEvents(candidate)
```

Comment:
Method 1 for triggering epoch fallback: we incorporate a block applying an invalid service event. Even if the block is not finalized, we trigger epoch fallback immediately.

Reply:
Created issue: https://github.com/dapperlabs/flow-go/issues/6316